Auth0/OAuth2Grant.swift

```diff
@@ -67,50 +81,99 @@ struct PKCE: OAuth2Grant {
             "code_challenge": challenge,
             "code_challenge_method": method
         ]
-
         if let nonce = nonce {
             newDefaults["nonce"] = nonce
         }
-
         self.defaults = newDefaults
     }
 
     func credentials(from values: [String: String], callback: @escaping (Result<Credentials>) -> Void) {
-        guard
-            let code = values["code"]
-            else {
-                let string = "No code found in parameters \(values)"
-                return callback(.failure(error: AuthenticationError(string: string)))
+        guard let code = values["code"] else {
+            let string = "No code found in parameters \(values)"
+            return callback(.failure(error: AuthenticationError(string: string)))
         }
-        guard validate(responseType: self.responseType, token: values["id_token"], nonce: self.defaults["nonce"]) else {
-            return callback(.failure(error: WebAuthError.invalidIdTokenNonce))
-        }
-        let clientId = self.authentication.clientId
-        self.authentication
-            .tokenExchange(withCode: code, codeVerifier: verifier, redirectURI: redirectURL.absoluteString)
-            .start { result in
-                // Special case for PKCE when the correct method for token endpoint authentication is not set (it should be None)
-                if case .failure(let cause as AuthenticationError) = result, cause.description == "Unauthorized" {
-                    let error = WebAuthError.pkceNotAllowed("Unable to complete authentication with PKCE. PKCE support can be enabled by setting Application Type to 'Native' and Token Endpoint Authentication Method to 'None' for this app at 'https://manage.auth0.com/#/applications/\(clientId)/settings'.")
-                    callback(Result.failure(error: error))
-                } else {
-                    callback(result)
-                }
+        let idToken = values["id_token"]
+        let responseType = self.responseType
+        let authentication = self.authentication
+        let leeway = self.leeway
+        let nonce = self.defaults["nonce"]
+        let maxAge = self.maxAge
+        let isHybridFlow = responseType.contains(.idToken)
+        let verifier = self.verifier
+        let redirectUrlString = self.redirectURL.absoluteString
+        let clientId = authentication.clientId
+        let validatorContext = IDTokenValidatorContext(authentication: authentication,
+                                                       leeway: leeway,
+                                                       nonce: nonce,
+                                                       maxAge: maxAge)
+        validate(idToken: idToken, for: responseType, with: validatorContext) { error in
+            if let error = error { return callback(.failure(error: error)) }
+            authentication
+                .tokenExchange(withCode: code, codeVerifier: verifier, redirectURI: redirectUrlString)
```

Comment:
I assume `tokenExchange` is an existing method? Since we will/do have token exchange in this library, `codeExchange` or similar would be more helpful/accurate.

Reply:
Yes, it's an existing (and public) method. Changing it would be a breaking change.